### PR TITLE
AEROGEAR-2454 AuthService and Example App implementations

### PR DESCRIPTION
**Pull request by aidenkeating**
## Motivation

JIRA: https://issues.jboss.org/browse/AEROGEAR-2454

## Description

AuthService implementations for Android and iOS.
Example app implementations for Android and iOS.

## Progress

- [x] AuthService implementations
- [x] Example App implementations
- [x] Documentation

### CODE REVIEW

File: Core/Core/MobileCore.cs [hunk 1]
```diff
@@ -196,14 +196,8 @@ private T GetInstance<T>(Type serviceClass, ServiceConfiguration serviceConfigur
...
-
-            IServiceModule serviceModule = Activator.CreateInstance(serviceClass) as IServiceModule;
-            var serviceCfg = serviceConfiguration;
-            if (serviceCfg == null) serviceCfg = GetServiceConfiguration(serviceModule.Type);
-            if (serviceCfg == null && serviceModule.RequiresConfiguration) throw new ConfigurationNotFoundException($"{serviceModule.Type} not found on " + ConfigFileName);
-            serviceModule.Configure(this, serviceCfg);
-            services[serviceClass] = serviceModule;
-            return (T)serviceModule;
+            // There are no services registered for this interface.
+            throw new ServiceModuleInstanceNotFoundException(serviceClass.Name);
```
Comment (aidenkeating) [inline Core/Core/MobileCore.cs]: @ziccardi Does this make sense to you? I was seeing the `Could not create instance of IAuthService` error before I registered the AuthService implementation. I think this error is more descriptive

**Comment by aidenkeating** [comment]
@wei-lee @ziccardi Mind taking a look?

File: Core/Core/MobileCore.cs (hunk 1 above)
Comment (ziccardi) [reply Core/Core/MobileCore.cs]: Yes, I would even add a message like: String.Format("No instance has been registered for interface {0}", serviceClass.Name)

**Comment by wei-lee** [comment]
The code changes looks good. I have ran both android and ios apps, found a few issues:

1. The android app doesn't seem to work with emulators that are running android API level 26.
2. For the ios app, for some reason, after the logout button is clicked, it took a while for the view to be updated.

I don't think the issues are related to the implementation in this PR here. So i am happy to merge this PR, and then we can fix those issues in separate PRs.

Good job! 👍

File: Auth/Auth/AbstractAuthService.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace AeroGear.Mobile.Auth
+{
+    /// <summary>
+    /// Abstract implementation of <see cref="IAuthService"/>.
+    /// </summary>
+    public abstract class AbstractAuthService : IAuthService
+    {
+        protected CredentialManager CredentialManager { get; set; }
+        protected KeycloakConfig KeycloakConfig { get; set; }
```
Comment (ziccardi) [inline Auth/Auth/AbstractAuthService.cs]: this should be readonly
```c#
protected readonly KeycloakConfig KeycloakConfig;
```

File: Auth/Auth/AbstractAuthService.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        protected AuthenticationConfig AuthenticationConfig { get; private set; }
+        protected MobileCore MobileCore { get; set; }
```
Comment (ziccardi) [inline Auth/Auth/AbstractAuthService.cs]: this should be readonly

**Comment by ziccardi** [comment]
@aidenkeating : was reviewing 😄 
As a rule of thumb, if a property is meant to be initialized only inside the constructor, it should be declared `readonly`

**Pull request merged**